File: src/issues.ts
/** 解析・検証で検出した問題の深刻度 */
export type IssueLevel = "error" | "warning";

export const ISSUE_CODE = {
  emptyInput: "empty-input",
  missingVersionRecord: "missing-version-record",
  unknownVersion: "unknown-version",
  unknownRecord: "unknown-record",
  duplicateRecord: "duplicate-record",
  missingRequiredField: "missing-required-field",
  tooManyFields: "too-many-fields",
  fieldTooLong: "field-too-long",
  invalidCode: "invalid-code",
  invalidDate: "invalid-date",
  invalidType: "invalid-type",
  orphanRecord: "orphan-record",
  missingUsage: "missing-usage",
  splitMismatch: "split-mismatch",
} as const;

export type IssueCode = (typeof ISSUE_CODE)[keyof typeof ISSUE_CODE];

export interface Issue {
  readonly level: IssueLevel;
  readonly code: IssueCode;
  readonly message: string;
  /** 1始まりの行番号 */
  readonly line?: number;
  readonly recordNo?: string;
  /** レコード内の項目位置(レコード番号を 0 とした連番) */
  readonly fieldIndex?: number;
  readonly fieldKey?: string;
}

export class IssueCollector {
  private readonly issues: Issue[] = [];

  add(issue: Issue): void {
    this.issues.push(issue);
  }

  error(code: IssueCode, message: string, context: Omit<Issue, "level" | "code" | "message"> = {}): void {
    this.add({ level: "error", code, message, ...context });
  }

  warn(code: IssueCode, message: string, context: Omit<Issue, "level" | "code" | "message"> = {}): void {
    this.add({ level: "warning", code, message, ...context });
  }

  get all(): Issue[] {
    return this.issues;
  }

  get hasError(): boolean {
    return this.issues.some((issue) => issue.level === "error");
  }
}

/** 構造化データをテキストへ変換する際のエラー */
export class SerializeError extends Error {
  constructor(
    message: string,
    readonly context?: { recordNo?: string; fieldKey?: string; value?: string },
  ) {
    super(message);
    this.name = "SerializeError";
  }
}

/** QRコードの読み取り・生成に失敗した際のエラー */
export class QrCodeError extends Error {
  constructor(message: string, options?: unknown) {
    super(message);
    this.name = "QrCodeError";
    if (options !== undefined) this.cause = options;
  }
}

File: python/src/medication_record_qr/issues.py
"""解析・検証で検出した問題と、ライブラリが送出する例外。"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Literal

IssueLevel = Literal["error", "warning"]


class IssueCode:
    """問題の種別。"""

    EMPTY_INPUT = "empty-input"
    MISSING_VERSION_RECORD = "missing-version-record"
    UNKNOWN_VERSION = "unknown-version"
    UNKNOWN_RECORD = "unknown-record"
    DUPLICATE_RECORD = "duplicate-record"
    MISSING_REQUIRED_FIELD = "missing-required-field"
    TOO_MANY_FIELDS = "too-many-fields"
    FIELD_TOO_LONG = "field-too-long"
    INVALID_CODE = "invalid-code"
    INVALID_DATE = "invalid-date"
    INVALID_TYPE = "invalid-type"
    ORPHAN_RECORD = "orphan-record"
    MISSING_USAGE = "missing-usage"
    SPLIT_MISMATCH = "split-mismatch"


@dataclass(frozen=True)
class Issue:
    """解析・検証で検出した1件の問題。"""

    level: IssueLevel
    code: str
    message: str
    #: 1始まりの行番号
    line: int | None = None
    record_no: str | None = None
    #: レコード内の項目位置(レコード番号を 0 とした連番)
    field_index: int | None = None
    field_key: str | None = None


class IssueCollector:
    """問題を集めるための入れ物。"""

    def __init__(self) -> None:
        self._issues: list[Issue] = []

    def add(self, issue: Issue) -> None:
        self._issues.append(issue)

    def error(self, code: str, message: str, **context: object) -> None:
        self.add(Issue(level="error", code=code, message=message, **context))  # type: ignore[arg-type]

    def warn(self, code: str, message: str, **context: object) -> None:
        self.add(Issue(level="warning", code=code, message=message, **context))  # type: ignore[arg-type]

    @property
    def all(self) -> list[Issue]:
        return self._issues

    @property
    def has_error(self) -> bool:
        return any(issue.level == "error" for issue in self._issues)


class MedicationNotebookError(Exception):
    """このライブラリが送出する例外の基底クラス。"""


class ParseError(MedicationNotebookError):
    """エラーを含むデータを ``parse_or_raise`` で解析した場合に送出される。"""

    def __init__(self, message: str, issues: list[Issue]) -> None:
        super().__init__(message)
        self.issues = issues


class SerializeError(MedicationNotebookError):
    """構造化データをテキストへ変換できない場合に送出される。"""

    def __init__(self, message: str, *, record_no: str | None = None, value: str | None = None) -> None:
        super().__init__(message)
        self.record_no = record_no
        self.value = value


class QrCodeError(MedicationNotebookError):
    """QRコードの生成・読み取りに失敗した場合に送出される。"""
